pkg/operator/controller/ingress/controller.go

```diff
@@ -713,6 +726,121 @@ func (r *reconciler) ensureIngressDeleted(ingress *operatorv1.IngressController)
+		if cleared {
+			routesCleared++
+		}
+		if err != nil {
+			return err
+		}
+	}
+	elapsed := time.Since(start)
+	log.Info("cleared route status for ingress deletion", "Ingress Controller", icName, "Routes Status Cleared", routesCleared, "Time Elapsed", elapsed)
+	return nil
+}
+
+// Clears this route's status that is admitted by provided ingressController
+func (r *reconciler) clearRouteStatus(route routev1.Route, icName string) (bool, error) {
+	// Go through each route and clear status if admitted by this ingress controller.
+	var updated routev1.Route
+	clearedStatus := false
+	for i := range route.Status.Ingress {
+		if condition := findCondition(&route.Status.Ingress[i], routev1.RouteAdmitted); condition != nil {
+			if routerName := &route.Status.Ingress[i].RouterName; routerName != nil && *routerName == icName {
+				// Remove this status since it matches our routerName.
+				route.DeepCopyInto(&updated)
+				updated.Status.Ingress = append(route.Status.Ingress[:i], route.Status.Ingress[i+1:]...)
+				if err := r.client.Status().Update(context.TODO(), &updated); err != nil {
+					return clearedStatus, fmt.Errorf("failed to clear route status for %s with routerName %s: %v", route.Name, icName, err)
+				}
+				clearedStatus = true
+			}
+		}
+	}
+
+	return clearedStatus, nil
+}
+
+// Clear routes that are not owned by this ingress controller
+func (r *reconciler) clearRoutesNotAdmittedByIngress(ingress *operatorv1.IngressController) {
+	// TODO: THIS IS TESTING CODE. NEEDS A LOT OF WORK
+	start := time.Now()
+
+	routesCleared := 0
+	// List all routes.
+	routeList := &routev1.RouteList{}
+	if err := r.cache.List(context.TODO(), routeList); err != nil {
+		fmt.Errorf("failed to list all routes in order to clear route status")
+	}
+	for _, route := range routeList.Items {
+		routeSelectorMatch := false
+		namespaceSelectorMatch := false
+
+		// TODO: REAL MESSY, FIX THIS UP
+		routeSelector, _ := metav1.LabelSelectorAsSelector(ingress.Spec.RouteSelector)
+		if routeSelector != nil {
+			routeLabelSelector, _ := labels.Parse(routeSelector.String())
+			if routeLabelSelector == nil {
+				routeSelectorMatch = true
+			} else if routeLabelSelector.Matches(labels.Set(route.Labels)) {
+				routeSelectorMatch = true
+			}
+		} else {
+			routeSelectorMatch = true
+		}
+
+		// TODO: What if namespace Expression is invalid? This will panic
+		namespaceSelector, _ := metav1.LabelSelectorAsSelector(ingress.Spec.NamespaceSelector)
+
+		namespaceLabelSelector, _ := labels.Parse(namespaceSelector.String())
+		ns := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: route.Namespace}}
+		if err := r.cache.Get(context.TODO(), types.NamespacedName{Name: route.Namespace}, ns); err != nil {
+			fmt.Errorf("failed to get namespace of route")
+		}
+		if namespaceLabelSelector == nil {
+			namespaceSelectorMatch = true
+		} else if namespaceLabelSelector.Matches(labels.Set(ns.Labels)) {
+			namespaceSelectorMatch = true
+		}
```

Comment:
Phooey, GitHub lost a comment that I meant to include in my review, so I gotta retype it...

What OpenShift router does is list/watch namespaces using the namespace selector and maintain a list of namespaces that it's supposed to handle, and then each time the router processes a route, it checks whether the route is in that list.  You could do something similar:

1. Create a string set (`namespaces := sets.NewString()`).
2. List namespaces using the namespace selector, similar to what I suggested for listing routes using the route selector.
3. Iterate over the namespaces, adding them to this string set (`namespaces.Insert(namespace.Name)`).  

Then use that set when you iterate over routes (`if !namespaces.Has(route.Namespace) { continue }`).

Reply:
> Then use that set when you iterate over routes (`if !namespaces.Has(route.Namespace) { continue }`).

Errr, actually you want to clear the status in this case:
```go
if !namespaces.Has(route.Namespace) {
	r.clearRouteStatus(route, ingress.ObjectMeta.Name)
}
```

Reply:
done